### PR TITLE
C51-269: Add Activity.getmonthswithactivities endpoint

### PR DESCRIPTION
**Pull request by 16kilobyte**
The PR adds `Activity.getmonthswithactivities` endpoint which returns a list of the map of unique `month` and `year` with at least one activity. The endpoint supports all the parameters that `Activity.get` does.

Sample response
```json
[
        {
            "year": "2009",
            "month": "1"
        },
        {
            "year": "2009",
            "month": "10"
        },
        {
            "year": "2010",
            "month": "3"
        },
        {
            "year": "2015",
            "month": "5"
        },
        {
            "year": "2015",
            "month": "12"
        },
        {
            "year": "2016",
            "month": "12"
        },
        {
            "year": "2017",
            "month": "1"
        },
        {
            "year": "2017",
            "month": "10"
        },
        {
            "year": "2017",
            "month": "11"
        },
        {
            "year": "2017",
            "month": "12"
        },
        {
            "year": "2018",
            "month": "11"
        }
]
```

### CODE REVIEW

File: api/v3/Activity/Getmonthswithactivities.php [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+/**
+ * Activity.Getmonthswithactivities API specification
+ *
+ * @param array $spec description of fields supported by this API call
+ * 
+ * @return void
+ * @see http://wiki.civicrm.org/confluence/display/CRMDOC/API+Architecture+Standards
+ */
+function _civicrm_api3_activity_Getmonthswithactivities_spec(&$spec) {
+  $allowed = ['activity_date_time', 'activity_status_id', 'case_id'];
```
Comment (deb1990) [inline api/v3/Activity/Getmonthswithactivities.php]: Why are we only allowing 3 params? It was never mentioned in the requirements.
The API should allow all parameters as `Activity.get`

File: civicase.civix.php [hunk 2]
```diff
@@ -443,3 +443,22 @@ function _civicase_civix_civicrm_alterSettingsFolders(&$metaDataFolders = NULL)
...
+
+/**
+ * (Delegated) Implements hook_civicrm_entityTypes().
+ *
+ * Find any *.entityType.php files, merge their content, and return.
+ *
+ * @link http://wiki.civicrm.org/confluence/display/CRMDOC/hook_civicrm_entityTypes
+ */
+
+function _civicase_civix_civicrm_entityTypes(&$entityTypes) {
```
Comment (deb1990) [inline civicase.civix.php]: Where is this function used? And why do we need it?
This looks completely unrelated.

File: api/v3/Activity/Getmonthswithactivities.php [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+ * @method Activity.Getmonthswithactivities API
+ *
+ * @param array $params
+ * @return array API result with list of months
+ * @see civicrm_api3_create_success
+ * @throws API_Exception
+ */
+function civicrm_api3_activity_Getmonthswithactivities($params) {
+  try {
+      $dao = CRM_Utils_SQL_Select::from('civicrm_activity a');
```
Comment (deb1990) [inline api/v3/Activity/Getmonthswithactivities.php]: Wrong indentation.

File: api/v3/Activity/Getmonthswithactivities.php [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+ * 
+ * @method Activity.Getmonthswithactivities API
+ *
+ * @param array $params
+ * @return array API result with list of months
+ * @see civicrm_api3_create_success
+ * @throws API_Exception
+ */
+function civicrm_api3_activity_Getmonthswithactivities($params) {
+  try {
```
Comment (deb1990) [inline api/v3/Activity/Getmonthswithactivities.php]: Why do we need a `try/catch` block here? Cant we handle the errors manually.
For example, https://github.com/compucorp/uk.co.compucorp.civicase/blob/C51-269-getmonthswithactivities-endpoint/api/v3/Case/Getdetails.php#L233

File: tests/phpunit/api/v3/Activity/GetmonthswithactivitiesTest.php [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+ * This is a generic test class implemented with PHPUnit.
+ * @group headless
+ */
+class api_v3_Activity_GetmonthswithactivitiesTest extends \PHPUnit_Framework_TestCase implements HeadlessInterface, HookInterface, TransactionalInterface {
+
+  /**
+   * Civi\Test has many helpers, like install(), uninstall(), sql(), and sqlFile().
+   * See: https://github.com/civicrm/org.civicrm.testapalooza/blob/master/civi-test.md
+   */
+  public function setUpHeadless() {
```
Comment (deb1990) [inline tests/phpunit/api/v3/Activity/GetmonthswithactivitiesTest.php]: Please create a new `BaseHeadlessTest` Class to make it reusable.
For example
https://github.com/compucorp/civihr/blob/master/uk.co.compucorp.civicrm.hrleaveandabsences/tests/phpunit/BaseHeadlessTest.php#L6

File: tests/phpunit/api/v3/Activity/GetmonthswithactivitiesTest.php [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+    return \Civi\Test::headless()
+      ->installMe(__DIR__)
+      ->apply();
+  }
+
+  /**
+   * The setup() method is executed before the test is executed (optional).
+   */
+  public function setUp() {
```
Comment (deb1990) [inline tests/phpunit/api/v3/Activity/GetmonthswithactivitiesTest.php]: Nothing is happening in this method as of now, so please remove until we really use this.

File: tests/phpunit/api/v3/Activity/GetmonthswithactivitiesTest.php [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+    parent::setUp();
+  }
+
+  /**
+   * The tearDown() method is executed after the test was executed (optional)
+   * This can be used for cleanup.
+   */
+  public function tearDown() {
```
Comment (deb1990) [inline tests/phpunit/api/v3/Activity/GetmonthswithactivitiesTest.php]: Nothing is happening in this method as of now, so please remove until we really use this.

**Comment by deb1990** [comment]
thanks @16kilobyte . But 12 secs is not really acceptable. Ideally it should be under 1-2 seconds.

I can see that you have used `YEAR(a.activity_date_time) AS year, MONTH(a.activity_date_time) AS month` in your SQL, but in my experience this really slows down the performance.

What if we do this,
1. pass all the params to `Activity.get` call, and only return the `activity_date_time` field.
2. Manipulate the list of results in PHP to form this array of Month/Year.

could you try this and see if it improves the performance?

File: api/v3/Activity/Getmonthswithactivities.php [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+ */
+function civicrm_api3_activity_Getmonthswithactivities($params)
+{
+    $result = civicrm_api3('Activity', 'get', array_merge($params, [
+      'sequential' => 1,
+      'return' => 'activity_date_time',
+      'options' => ['limit' => 0],
+    ]));
+
+    if (!boolval($result['is_error'])) {
```
Comment (reneolivo) [inline api/v3/Activity/Getmonthswithactivities.php]: I think it's better to fail fast and also avoid having the majority of the code inside an if block. Lets try this instead:

```php
if (boolval($result['is_error'])) {
  return civicrm_api3_create_error($result['error_message'], $params);
}

$grouped_activity_dates = array();
$grouped_activity_dates_indexes = array();
$index = 0;

// rest of the code
```

File: api/v3/Activity/Getmonthswithactivities.php [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+      $grouped_activity_dates = array();
+      $grouped_activity_dates_indexes = array();
+      $index = 0;
+
+      foreach($result['values'] as $activity_date_time) {
+        list($activity_year, $activity_month) = explode('-', $activity_date_time['activity_date_time']);
+        
+        if (!isset($grouped_activity_dates_indexes[$activity_month . $activity_year])) {
+          $grouped_activity_dates_indexes[$activity_month . $activity_year] = $index;
+          $index++;
```
Comment (reneolivo) [inline api/v3/Activity/Getmonthswithactivities.php]: I don't think we need the `$index` variable, you could just assign `true`:

```php

foreach($result['values'] as $activity_date_time) {
  list($activity_year, $activity_month) = explode('-', $activity_date_time['activity_date_time']);

  if (!isset($grouped_activity_dates_indexes[$activity_month . $activity_year])) {
    $grouped_activity_dates_indexes[$activity_month . $activity_year] = true;

    $grouped_activity_dates[] = // ... rest of the code
```

File: api/v3/Activity/Getmonthswithactivities.php (hunk 1 above)
Comment (16kilobyte) [reply api/v3/Activity/Getmonthswithactivities.php]: Endpoint now allows all of `Activity.get`'s parameters

File: api/v3/Activity/Getmonthswithactivities.php (hunk 9 above)
Comment (16kilobyte) [reply api/v3/Activity/Getmonthswithactivities.php]: I kept the `$index` variable in case there's need to order the returned data

Comment (16kilobyte) [reply api/v3/Activity/Getmonthswithactivities.php]: Right. I kept `$index` so that data could be sorted if there's need.

File: civicase.civix.php (hunk 2 above)
Comment (16kilobyte) [reply civicase.civix.php]: Function was auto generated. I have removed it.
